jolt-core/src/jolt/vm/bytecode.rs

```diff
@@ -492,160 +840,298 @@ pub fn prove_booleanity<F: JoltField, ProofTranscript: Transcript>(
+                || vec![F::zero(); 3],
+                |mut running, new| {
+                    for i in 0..3 {
+                        running[i] += new[i];
+                    }
+                    running
                 },
             );
 
-        univariate_poly_evals = [
-            eq_r_r * univariate_poly_evals[0],
-            eq_r_r * univariate_poly_evals[1],
-            eq_r_r * univariate_poly_evals[2],
-        ];
+        // Multiply by eq_r_r
+        for eval in &mut univariate_poly_evals {
+            *eval *= prover_state.eq_r_r;
+        }
 
-        let univariate_poly = UniPoly::from_evals(&[
-            univariate_poly_evals[0],
-            previous_claim - univariate_poly_evals[0],
-            univariate_poly_evals[1],
-            univariate_poly_evals[2],
-        ]);
+        univariate_poly_evals
+    }
+}
 
-        drop(_inner_guard);
-        drop(inner_span);
+#[derive(CanonicalSerialize, CanonicalDeserialize, Debug, Clone)]
+pub struct BooleanityProof<F: JoltField, ProofTranscript: Transcript> {
+    sumcheck_proof: SumcheckInstanceProof<F, ProofTranscript>,
+    ra_claim_prime: F,
+}
 
-        let compressed_poly = univariate_poly.compress();
-        compressed_poly.append_to_transcript(transcript);
-        compressed_polys.push(compressed_poly);
+impl<F: JoltField, ProofTranscript: Transcript> BooleanityProof<F, ProofTranscript> {
+    #[tracing::instrument(skip_all, name = "BooleanityProof::prove")]
+    pub fn prove(
+        preprocessing: &BytecodePreprocessing,
+        trace: &[RV32IMCycle],
+        r: &[F],
+        D: Vec<F>,
+        G: Vec<F>,
+        transcript: &mut ProofTranscript,
+    ) -> (Self, Vec<F>, Vec<F>) {
+        let K = r.len().pow2();
+        let T = trace.len();
 
-        let r_j = transcript.challenge_scalar::<F>();
-        r_cycle_prime.push(r_j);
+        let mut booleanity_sumcheck = BooleanitySumcheck::new(preprocessing, trace, r, D, G, K, T);
 
-        previous_claim = univariate_poly.evaluate(&r_j);
+        let (sumcheck_proof, r_combined) = booleanity_sumcheck.prove_single(transcript);
+
+        let (r_address_prime, r_cycle_prime) = r_combined.split_at(K.log_2());
+
+        let ra_claim_prime = booleanity_sumcheck
+            .ra_claim_prime
+            .expect("ra_claim_prime should be set after prove_single");
+
+        let proof = Self {
+            sumcheck_proof,
+            ra_claim_prime,
+        };
+
+        (proof, r_address_prime.to_vec(), r_cycle_prime.to_vec())
+    }
+
+    pub fn verify(
+        &self,
+        r_address: &[F],
+        r_cycle: &[F],
+        K: usize,
+        T: usize,
+        transcript: &mut ProofTranscript,
+    ) -> Result<(Vec<F>, F), ProofVerifyError> {
+        let mut booleanity_sumcheck = BooleanitySumcheck::new_verifier(K, T);
+
+        booleanity_sumcheck.set_r_values(r_address.to_vec(), r_cycle.to_vec());
+
+        booleanity_sumcheck.ra_claim_prime = Some(self.ra_claim_prime);
```

Comment:
we should just pass in `r_address`, `r_cycle`, and `self.ra_claim_prime` to `BooleanitySumcheck::new_verifier` imo